library-best-practices/JLBP-15.md

```diff
@@ -0,0 +1,16 @@
+[JLBP-15] Avoid dependency and project cycles
+---------------------------------------------
+
+- Avoid cycles in artifact dependencies. If there is an initial temptation to
+  add a cycle, instead lift up any shared types between the relevant artifacts
+  into a new shared artifact.
+- Avoid cycles in project dependencies (even if it doesn't technically create a
+  cycle between artifacts).
+  - Example violation: grpc-core depends on opencensus-api which depends on
+    grpc-context, and grpc-core and grpc-context are always released together
```

Comment:
The dependency chain is grpc-core -> opencensus-api -> grpc-context. Also, grpc-core depends on grpc-context directly. 

It is not a "dependency" cycle, it is a "project" cycle. You are correct that it is not what is usually meant by a cycle. We do talk about basic cycles first - that is the first bullet.

I'm not sure who should "blink" first; maybe grpc-context should be released separately from the rest of grpc, because maybe other aspectual features might need to reference it too.

Reply:
What do you think about splitting this into separate JLBPs?

Reply:
If Maven fails on artifact cycles, do we even need to mention it at all? We could just scope this JLBP to project cycles and drop the mention of artifact cycles altogether.